File: apps/voicememory_mobile/lib/features/archive_maturity/archive_maturity_engine.dart
import '../../models/journal_entry.dart';

enum ArchiveMaturityStage {
  starting,
  buildingEvidence,
  beliefsForming,
  beliefsChanging,
  matureArchive,
}

class ArchiveMaturityInput {
  const ArchiveMaturityInput({
    required this.reflectionCount,
    required this.beliefCount,
    required this.beliefChanges,
    required this.reputationScore,
    required this.timelineAgeDays,
  });

  final int reflectionCount;
  final int beliefCount;
  final int beliefChanges;
  final int reputationScore;
  final int timelineAgeDays;
}

class ArchiveProgressView {
  const ArchiveProgressView({
    required this.score,
    required this.stage,
    required this.stageLabel,
    required this.headline,
    required this.nextMilestoneLabel,
    required this.nextMilestonePercent,
  });

  final int score;
  final ArchiveMaturityStage stage;
  final String stageLabel;
  final String headline;
  final String nextMilestoneLabel;
  final int nextMilestonePercent;
}

/// Unified archive maturity score (0–100) for mobile progress bar.
class ArchiveMaturityEngine {
  static const headline = 'Your archive is becoming harder to fool.';

  static const _milestones = <({int percent, String label})>[
    (percent: 20, label: 'Enough reflections to compare'),
    (percent: 40, label: 'First beliefs under review'),
    (percent: 60, label: 'Beliefs shifting with new evidence'),
    (percent: 80, label: 'Archive reputation strengthening'),
    (percent: 100, label: 'Mature archive — beliefs hard to fool'),
  ];

  static int compute(ArchiveMaturityInput input) {
    final raw =
        (input.reflectionCount.clamp(0, 15) * 3.2) +
        (input.beliefCount.clamp(0, 8) * 5.5) +
        (input.beliefChanges.clamp(0, 12) * 4) +
        (input.reputationScore.clamp(0, 100) * 0.28) +
        (input.timelineAgeDays.clamp(0, 120) * 0.12);
    return raw.round().clamp(0, 100);
  }

  static ArchiveMaturityStage stageFor({
    required int reflectionCount,
    required int beliefCount,
    required int beliefChanges,
    required int score,
  }) {
    if (reflectionCount <= 1) return ArchiveMaturityStage.starting;
    if (reflectionCount <= 3 && beliefCount == 0) {
      return ArchiveMaturityStage.buildingEvidence;
    }
    if (beliefCount == 0) return ArchiveMaturityStage.buildingEvidence;
    if (beliefChanges >= 2 || score >= 72) {
      return ArchiveMaturityStage.matureArchive;
    }
    if (beliefChanges >= 1 || score >= 48) {
      return ArchiveMaturityStage.beliefsChanging;
    }
    return ArchiveMaturityStage.beliefsForming;
  }

  static String stageLabel(ArchiveMaturityStage stage) {
    switch (stage) {
      case ArchiveMaturityStage.starting:
        return 'Starting';
      case ArchiveMaturityStage.buildingEvidence:
        return 'Building evidence';
      case ArchiveMaturityStage.beliefsForming:
        return 'Beliefs forming';
      case ArchiveMaturityStage.beliefsChanging:
        return 'Beliefs changing';
      case ArchiveMaturityStage.matureArchive:
        return 'Mature archive';
    }
  }

  static ({String label, int percent}) nextMilestone(int score) {
    for (final m in _milestones) {
      if (m.percent > score) return (label: m.label, percent: m.percent);
    }
    return _milestones.last;
  }

  static ArchiveProgressView buildView(ArchiveMaturityInput input) {
    final score = compute(input);
    final stage = stageFor(
      reflectionCount: input.reflectionCount,
      beliefCount: input.beliefCount,
      beliefChanges: input.beliefChanges,
      score: score,
    );
    final milestone = nextMilestone(score);
    return ArchiveProgressView(
      score: score,
      stage: stage,
      stageLabel: stageLabel(stage),
      headline: headline,
      nextMilestoneLabel: milestone.label,
      nextMilestonePercent: milestone.percent,
    );
  }

  static ArchiveMaturityInput inputFromEntries(List<JournalEntry> entries) {
    final eligible = entries
        .where((e) => e.transcript.trim().isNotEmpty)
        .toList();
    eligible.sort((a, b) => a.createdAt.compareTo(b.createdAt));

    final reflectionCount = eligible.length;
    final beliefCount = reflectionCount >= 5 ? 1 : 0;
    final beliefChanges = reflectionCount >= 6 ? 1 : 0;

    int timelineDays = 0;
    if (eligible.isNotEmpty) {
      timelineDays =
          eligible.last.createdAt
              .difference(eligible.first.createdAt)
              .inDays
              .abs() +
          1;
    }

    final reputationScore = (reflectionCount * 8 + timelineDays).clamp(0, 100);

    return ArchiveMaturityInput(
      reflectionCount: reflectionCount,
      beliefCount: beliefCount,
      beliefChanges: beliefChanges,
      reputationScore: reputationScore,
      timelineAgeDays: timelineDays,
    );
  }
}

File: packages/shared/lib/archive/archive-maturity-engine.ts
import { daysBetweenKeys, toDayKey } from "@/lib/dates";
import { readBeliefTimelineHistory } from "@/lib/archive/belief-timeline-storage";
import {
  ARCHIVE_MATURITY_STAGE_LABEL,
} from "@/lib/archive/archive-maturity-copy";
import {
  ARCHIVE_PROGRESS_HEADLINE,
  nextMilestoneForScore,
} from "@/lib/archive/archive-progress-copy";
import { buildArchiveReputationView } from "@/lib/archive/archive-reputation";
import { buildArchiveValueSnapshot } from "@/lib/product/archive-value-progress";
import { buildTheoryTrackerReport } from "@/lib/theories/theory-generation";
import { getMemoryEligibleEntries } from "@/lib/storage";
import type { ArchiveMaturityStage } from "@/types/archive-maturity";
import type {
  ArchiveMaturityEngineInput,
  ArchiveProgressView,
} from "@/types/archive-progress";
import type { JournalEntry } from "@/types/journal";

function eligible(entries: JournalEntry[]): JournalEntry[] {
  return entries.filter((e) => e.reflectionPending !== true);
}

function timelineAgeDays(entries: JournalEntry[]): number {
  if (entries.length === 0) return 0;
  if (entries.length === 1) return 1;
  const sorted = [...entries].sort(
    (a, b) => new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime(),
  );
  const first = toDayKey(sorted[0]!.createdAt);
  const last = toDayKey(sorted[sorted.length - 1]!.createdAt);
  return Math.max(1, daysBetweenKeys(first, last) + 1);
}

function beliefChangeCount(entries: JournalEntry[], theoryIds: string[]): number {
  let total = 0;
  for (const id of theoryIds) {
    total += readBeliefTimelineHistory(id).length;
  }
  if (total > 0) return total;
  const report = buildTheoryTrackerReport(entries, { persistSnapshots: false });
  return report.all.filter((t) => (t.whatChanged?.length ?? 0) > 0).length;
}

function stageFromSignals(signals: {
  reflectionCount: number;
  beliefCount: number;
  beliefChanges: number;
  score: number;
}): ArchiveMaturityStage {
  const { reflectionCount, beliefCount, beliefChanges, score } = signals;
  if (reflectionCount <= 1) return "starting";
  if (reflectionCount <= 3 && beliefCount === 0) return "building_evidence";
  if (beliefCount === 0) return "building_evidence";
  if (beliefChanges >= 2 || score >= 72) return "mature_archive";
  if (beliefChanges >= 1 || score >= 48) return "beliefs_changing";
  return "beliefs_forming";
}

/** Unified archive maturity score from five signals (0–100). */
export class ArchiveMaturityEngine {
  static compute(input: ArchiveMaturityEngineInput): number {
    const raw =
      Math.min(input.reflectionCount, 15) * 3.2 +
      Math.min(input.beliefCount, 8) * 5.5 +
      Math.min(input.beliefChanges, 12) * 4 +
      Math.min(input.reputationScore, 100) * 0.28 +
      Math.min(input.timelineAgeDays, 120) * 0.12;

    return Math.min(100, Math.max(0, Math.round(raw)));
  }

  static buildView(input: ArchiveMaturityEngineInput): ArchiveProgressView {
    const score = ArchiveMaturityEngine.compute(input);
    const stage = stageFromSignals({
      reflectionCount: input.reflectionCount,
      beliefCount: input.beliefCount,
      beliefChanges: input.beliefChanges,
      score,
    });
    const milestone = nextMilestoneForScore(score);

    return {
      score,
      stage,
      stageLabel: ARCHIVE_MATURITY_STAGE_LABEL[stage],
      headline: ARCHIVE_PROGRESS_HEADLINE,
      nextMilestoneLabel: milestone.label,
      nextMilestonePercent: milestone.percent,
    };
  }
}

export function buildArchiveMaturityEngineInput(
  entriesInput?: JournalEntry[],
): ArchiveMaturityEngineInput {
  const entries = eligible(entriesInput ?? getMemoryEligibleEntries());
  const snapshot = buildArchiveValueSnapshot(entries);
  const report =
    entries.length >= 2
      ? buildTheoryTrackerReport(entries, { persistSnapshots: false })
      : null;

  const beliefCount =
    (report?.active.length ?? 0) + (report?.strengthening.length ?? 0);
  const theoryIds = report?.all.map((t) => t.id) ?? [];
  const reputation = buildArchiveReputationView(entries);

  return {
    reflectionCount: snapshot.reflectionCount,
    beliefCount,
    beliefChanges: beliefChangeCount(entries, theoryIds),
    reputationScore: reputation?.meterFill ?? 0,
    timelineAgeDays: timelineAgeDays(entries),
  };
}

export function buildArchiveProgressView(
  entriesInput?: JournalEntry[],
): ArchiveProgressView {
  return ArchiveMaturityEngine.buildView(buildArchiveMaturityEngineInput(entriesInput));
}
